storage/imx6ull-flash/flashdev.c

```diff
@@ -12,23 +12,102 @@
  */
 
 #include <errno.h>
+#include <limits.h>
+#include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
+
 #include <sys/mman.h>
 #include <sys/threads.h>
 #include <sys/minmax.h>
 
 #include "imx6ull-flashdev.h"
 #include "imx6ull-flashdrv.h"
 
-#define ECC_BITFLIP_THRESHOLD 10
+#define ECC_GF                13  /* Galois field */
+#define ECC_BITFLIP_THRESHOLD 10  /* Min number of bitflips for page rewrite */
+#define ECC0_BITFLIP_STRENGHT 16  /* ECC strength for metadata chunk */
+#define ECCN_BITFLIP_STRENGHT 14  /* ECC strength for data chunk */
+#define ECCN_DATA_SIZE        512 /* Data chunk size */
 
 
 /* Auxiliary functions */
 
-static int _flashmtd_checkECC(flashdrv_meta_t *meta, unsigned int chunks)
+static unsigned int _flashmtd_checkErased(const void *buff, size_t boffs, size_t blen)
 {
-	unsigned int i;
+	const unsigned char *buff8 = buff;
+	const unsigned int *buff32 = buff;
+	unsigned int data32, ret = 0;
+	unsigned char data8;
+
+	buff8 += boffs / CHAR_BIT;
+	boffs %= CHAR_BIT;
+
+	/* Check first byte */
+	if (boffs > 0) {
+		data8 = *buff8++;
+		data8 |= (unsigned char)(0xff << (CHAR_BIT - boffs));
+
+		if (boffs + blen < CHAR_BIT) {
+			data8 |= (unsigned char)(0xff >> (boffs + blen));
+			blen = 0;
+		}
+		else {
+			blen -= CHAR_BIT - boffs;
+		}
+
+		ret += CHAR_BIT - __builtin_popcount(data8);
+	}
+
+	/* Check bytes until 32-bit aligned address */
+	while ((blen > CHAR_BIT) && (((uintptr_t)buff8) % sizeof(unsigned int))) {
+		data8 = *buff8++;
+		blen -= CHAR_BIT;
+		ret += CHAR_BIT - __builtin_popcount(data8);
+	}
+
+	/* Check 32-bit words */
+	buff32 = (const unsigned int *)(uintptr_t)buff8;
```

Comment:
Is this about MISRA? Rules have been violated anyway (well, too bad, we'll justify that if we'll need to), this intermediate cast to uintptr_t shouldn't be needed